crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -1364,6 +1401,58 @@ def upgrade_compute_nodes(virt)
         compute_nodes.each do |n|
           upgrade_compute_node(controller, n)
         end
+      end
+
+      # Non-disruptive way of upgrading more compute nodes at once.
+      # Does live-evacuation of each compute node first, then starts parallel upgrade.
+      # Argument is list of node names.
+      def non_disruptive_upgrade_compute_nodes(names)
+        compute_nodes = names.map do |name|
+          ::Node.find_node_by_name_or_alias(name)
+        end
+
+        controller = fetch_nova_controller
+
+        # If there's a compute node which we already started to upgrade,
+        # (and the upgrade process was restarted due to the failure)
+        # continue with that one.
+        compute_nodes.sort! { |n| n.upgrading? ? -1 : 1 }
+
+        while compute_nodes.any?
+          # 1. Evacuate as many as possible, create a subset of nodes without instances.
+          nodes_to_upgrade = []
+          already_upgrading = false
+          compute_nodes.each do |node|
+            # if nodes are already in upgrading state, we can skip the live-migration part
+            already_upgrading = true if node.upgrading?
+            break if already_upgrading && !node.upgrading?
+            hostname = node[:hostname]
+            begin
+              live_evacuate_compute_node(controller, hostname) unless already_upgrading
```

Comment:
Isn't there the "risk" that this will move instances to a host that is also supposed to be upgraded within the current batch? I think the easiest way to address this to require the operator to manually disable and evacuate the compute nodes that are about to be upgraded.

Alternatively we could disable the whole batch of computes before actually starting the evacuation. To avoid instances being move around too many times.

Reply:
The risk is there, indeed. But as I stated elsewhere, this methods is only used for nodes that are explicitly selected by operator. So we could really require them (by documentation maybe) to manually disable (or even evacuate, but that's not necessary) the services before the action.

Reply:
There is also something to the other idea, yes. Potential problem is that the set might be too big and evacuation of all of the host might not be possible if all of them are disabled (same problem as in the discussion of the other comment)

Reply:
> There is also something to the other idea, yes. Potential problem is that the set might be too big and evacuation of all of the host might not be possible if all of them are disabled

This is also true for the first idea. That's why I suggested to make this the operator's "problem" (including the evacuation). He should know much better than our code what's going on.